staging/src/k8s.io/component-base/logs/json/json.go

```diff
@@ -0,0 +1,190 @@
+
+// Info write message to error level log
+func (l *infoLogger) Info(msg string, keysAndVals ...interface{}) {
+	if checkedEntry := l.l.Check(l.lvl, msg); checkedEntry != nil {
+		checkedEntry.Time = timeNow()
+		checkedEntry.Write(handleFields(l.l, keysAndVals)...)
+	}
+}
+
+// handleFields converts a bunch of arbitrary key-value pairs into Zap fields.  It takes
+// additional pre-converted Zap fields, for use with automatically attached fields, like
+// `error`.
+func handleFields(l *zap.Logger, args []interface{}, additional ...zap.Field) []zap.Field {
+	// a slightly modified version of zap.SugaredLogger.sweetenFields
+	if len(args) == 0 {
+		// fast-return if we have no suggared fields.
+		return additional
+	}
+
+	// unlike Zap, we can be pretty sure users aren't passing structured
+	// fields (since logr has no concept of that), so guess that we need a
+	// little less space.
+	fields := make([]zap.Field, 0, len(args)/2+len(additional))
+	for i := 0; i < len(args); {
+		// check just in case for strongly-typed Zap fields, which is illegal (since
+		// it breaks implementation agnosticism), so we can give a better error message.
+		if _, ok := args[i].(zap.Field); ok {
+			l.DPanic("strongly-typed Zap Field passed to logr", zap.Any("zap field", args[i]))
+			break
+		}
+
+		// make sure this isn't a mismatched key
+		if i == len(args)-1 {
+			l.DPanic("odd number of arguments passed as key-value pairs for logging", zap.Any("ignored key", args[i]))
+			break
+		}
+
+		// process a key-value pair,
+		// ensuring that the key is a string
+		key, val := args[i], args[i+1]
+		keyStr, isString := key.(string)
+		if !isString {
+			// if the key isn't a string, DPanic and stop logging
+			l.DPanic("non-string key argument passed to logging, ignoring all later arguments", zap.Any("invalid key", key))
+			break
+		}
+
+		fields = append(fields, zap.Any(keyStr, val))
+		i += 2
+	}
+
+	return append(fields, additional...)
+}
+
+// zapLogger is a logr.Logger that uses Zap to log.
+type zapLogger struct {
+	// NB: this looks very similar to zap.SugaredLogger, but
+	// deals with our desire to have multiple verbosity levels.
+	l *zap.Logger
+	infoLogger
+}
+
+var _ logr.Logger = &zapLogger{}
+
+// Error write log message to error level
+func (l *zapLogger) Error(err error, msg string, keysAndVals ...interface{}) {
+	if checkedEntry := l.l.Check(zap.ErrorLevel, msg); checkedEntry != nil {
+		checkedEntry.Time = timeNow()
+		checkedEntry.Write(handleFields(l.l, keysAndVals, handleError(err))...)
+	}
+}
+
+// V return info logr.Logger with specified level
+func (l *zapLogger) V(level int) logr.InfoLogger {
+	lvl := zapcore.Level(-1 * level)
+	if l.l.Core().Enabled(lvl) {
+		l.lvl = lvl
+		return l
+	}
+	return disabledInfoLogger
+}
+
+// WithValues return logr.Logger with some keys And Values
+func (l *zapLogger) WithValues(keysAndValues ...interface{}) logr.Logger {
+	newLogger := l.l.With(handleFields(l.l, keysAndValues)...)
```

Comment:
I expect that `WithValues` and `WithName` will be called frequently in future. You use `NewJsonLogger` which looks costly (allocates a lot of new objects on heap)